lib/irb.rb

```diff
@@ -1222,27 +1222,33 @@ def handle_exception(exc)
         irb_bug = true
       else
         irb_bug = false
-        # This is mostly to make IRB work nicely with Rails console's backtrace filtering, which patches WorkSpace#filter_backtrace
-        # In such use case, we want to filter the exception's backtrace before its displayed through Exception#full_message
-        # And we clone the exception object in order to avoid mutating the original exception
-        # TODO: introduce better API to expose exception backtrace externally
-        backtrace = exc.backtrace.map { |l| @context.workspace.filter_backtrace(l) }.compact
+        # To support backtrace filtering while utilizing Exception#full_message, we need to clone
+        # the exception to avoid modifying the original exception's backtrace.
         exc = exc.clone
-        exc.set_backtrace(backtrace)
-      end
+        filtered_backtrace = exc.backtrace.map { |l| @context.workspace.filter_backtrace(l) }.compact
+        backtrace_filter = IRB.conf[:BACKTRACE_FILTER]
 
-      if RUBY_VERSION < '3.0.0'
-        if STDOUT.tty?
-          message = exc.full_message(order: :bottom)
-          order = :bottom
-        else
-          message = exc.full_message(order: :top)
-          order = :top
+        if backtrace_filter
+          if backtrace_filter.respond_to?(:filter)
+            filtered_backtrace = backtrace_filter.filter(filtered_backtrace)
```

Comment:
I think `backtrace_filter.call` is better, assuming `backtrace_filter` to be a Proc. `.filter` is too specific to ActiveSupport::BacktraceCleaner's API. With `.call`, Rails can still specify backtrace_cleaner as a filter by:
```ruby
IRB.conf[:BACKTRACE_FILTER] = -> bt { backtrace_cleaner_instance.filter(bt) }
```
What do you think?

Reply:
👍 updated